### PR TITLE
gossip: init at unstable-2023-09-07 WIP

### PR DESCRIPTION
**Pull request by totoroot**
## Description of changes

I tried packaging `gossip`, a nostr desktop client.

I figured out all the necessary dependencies for it to compile, but ran into some issues when building.
Perhaps someone can help who knows more about packaging Rust applications than I do.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by totoroot** [comment]
Here is the log of the build failure:
```nix
warning: variable does not need to be mutable
    --> src/people.rs:1012:17
     |
1012 |             let mut person = elem.value_mut();
     |                 ----^^^^^^
     |                 |
     |                 help: remove this `mut`
     |
     = note: `#[warn(unused_mut)]` on by default

warning: variable does not need to be mutable
   --> src/people.rs:983:17
    |
983 |             let mut person = elem.value_mut();
    |                 ----^^^^^^
    |                 |
    |                 help: remove this `mut`

warning: variable does not need to be mutable
   --> src/ui/feed/note/mod.rs:717:9
    |
717 |     let mut style = ui.style_mut();
    |         ----^^^^^
    |         |
    |         help: remove this `mut`

warning: `gossip` (bin "gossip") generated 3 warnings (run `cargo fix --bin "gossip"` to apply 3 suggestions)
    Finished release [optimized] target(s) in 5m 28s
Executing cargoInstallPostBuildHook
cp: missing destination file operand after 'target/x86_64-unknown-linux-gnu/release-tmp/'
Try 'cp --help' for more information.
```

**Comment by matthiasbeyer** [comment]
I reported the build failure upstream in https://github.com/mikedilger/gossip/issues/489

Also... why was the `Cargo.lock` file committed here? Upstream has the lockfile committed in their repository, so there should be no need to commit it here, is there?

**Comment by matthiasbeyer** [comment]
Ok I am a bit at loss here. I updated the expression to the latest master of the project and still am getting that build failures I reported in the gossip issue.

Maybe it is just nixpkgs rustc being too old? I don't know. That's the last idea I would have...

File: pkgs/applications/networking/instant-messengers/gossip/default.nix [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    cargo build --release --all-features
+
+    runHook postBuild
+  '';
+
+  buildInputs = [
+    ffmpeg
+    openssl
+  ];
```
Comment (wegank) [inline pkgs/applications/networking/instant-messengers/gossip/default.nix:63]:
```suggestion
  ] ++ lib.optionals stdenv.isDarwin [
    darwin.apple_sdk_11_0.frameworks.IOKit
    ... (ofborg tells you)
  ];
```

Comment (totoroot) [reply pkgs/applications/networking/instant-messengers/gossip/default.nix:63]: While you're right, that's currently not the issue. It does not build properly for Linux either. Not sure, why ofborg is reporting it does.

**Comment by matthiasbeyer** [comment]
I took the liberty to push two patches to this PR which fix the build on linux for me. @totoroot please have a look!

It is not 0.7.0 anymore though, but latest master (as of the date of the commit). We might walk from there..

**Comment by matthiasbeyer** [comment]
It crashes at startup for me with:

```
thread 'main' panicked at 'Failed to initialize any backend! Wayland status: NoCompositorListening X11 status: LibraryOpenError(OpenError { kind: Library, deta
il: "opening library failed (libX11.so.6: cannot open shared object file: No such file or directory); opening library failed (libX11.so: cannot open shared obj
ect file: No such file or directory)" })', /build/cargo-vendor-dir/winit-0.28.6/src/platform_impl/linux/mod.rs:757:9                                           
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace   
```

**Comment by emmanuelrosa** [comment]
I have a working Nix package for Gossip 0.9.0, but it uses upstream's dep package rather than building from source; I didn't want to deal with Rust.

The underlying GUI system makes heavy use of dynamically-loaded libraries, which leads to the _opening library failed_ errors when the libraries can't be found. I'm using the `runtimeDependencies` attribute of `autoPatchelfHook` to make the necessary libraries available at runtime. See https://github.com/emmanuelrosa/erosanix/blob/master/pkgs/gossip/default.nix#L18

**Comment by emmanuelrosa** [comment]
I modified this package and got it to run. Video playpack works, but the audio doesn't seem to work. That's where I'm stuck.

This is for version 0.9, the latest release: https://github.com/emmanuelrosa/erosanix/tree/master/pkgs/gossip-full

Note: I'm using a custom Nix function not available in Nixpkgs to handle the .desktop icon. So that code will have to be replaced

gossip-screenshot

**Comment by totoroot** [comment]
@emmanuelrosa Since I do not plan on working on this any more, I will happily close this draft, if you decide to open a PR.
Please just let me know once you do and I will reference it here, so that people can easily find your proposed changes going forward.

**Comment by emmanuelrosa** [comment]
I've considered it @totoroot, but I won't be opening a Nixpkgs PR for Gossip.

The problem I keep running into with maintaining low-use/minimal-interest packages in Nixpkgs is that it takes a very long time to get version bump PRs merged. In short, it can take months.

While waiting for the PR to get merged, I take on additional overhead in order to use the updated package. My currrent strategy is to maintain a copy of the package in my own Nix flake and then use that copy instead of the one in Nixpkgs. So, from my point of view, why not just maintain packages in my own Nix flake and just forget about contributing such packages to Nixpkgs altogether?

I haven't seen enough interest in running Gossip on NixOS, to convince me that maintaining it in Nixpkgs is worth the effort.

The few who want to run Gossip on NixOS can use my package at https://github.com/emmanuelrosa/erosanix/tree/master/pkgs/gossip

That package is based on upstream's .deb package, and is therefore much easier for me to maintain than the one built from source.

**Comment by alyssais** [comment]
https://github.com/NixOS/nixpkgs/pull/299712

**Pull request closed without merging**